// @fye/netsuite-client/lib/netsuite-sales-tax.js
/* eslint-disable lines-between-class-members */
/* eslint-disable no-console */
// const debug = require('debug')('netsuite-sales-tax');

/**
 * NetSuiteSalesTax class
 */
class NetSuiteSalesTax {
  /**
   * @private  This constructor is for internal use only
   */
  constructor({
    taxDetailsReference,
    lineType,
    lineName,
    taxTypeInternalId,
    taxTypeName,
    taxCodeInternalId,
    taxCodeName,
    calcDetail,
    netAmount,
    taxBasis,
    taxRate,
    taxAmount,
  }) {
    this.taxDetailsReference = taxDetailsReference;
    this.lineType = lineType;
    this.lineName = lineName;
    this.taxTypeInternalId = taxTypeInternalId;
    this.taxTypeName = taxTypeName;
    this.taxCodeInternalId = taxCodeInternalId;
    this.taxCodeName = taxCodeName;
    this.calcDetail = calcDetail;
    this.netAmount = netAmount;
    this.taxBasis = taxBasis;
    this.taxRate = taxRate;
    this.taxAmount = taxAmount;
  }

  /**
   * @private  This method is for internal use only
   */
  equals(netsuiteSalesTax) {
    if (
      this.taxDetailsReference === netsuiteSalesTax.taxDetailsReference &&
      this.lineType === netsuiteSalesTax.lineType &&
      this.lineName === netsuiteSalesTax.lineName &&
      this.taxTypeInternalId === netsuiteSalesTax.taxTypeInternalId &&
      this.taxTypeName === netsuiteSalesTax.taxTypeName &&
      this.taxCodeInternalId === netsuiteSalesTax.taxCodeInternalId &&
      this.taxCodeName === netsuiteSalesTax.taxCodeName &&
      this.calcDetail === netsuiteSalesTax.calcDetail &&
      this.netAmount === netsuiteSalesTax.netAmount &&
      this.taxBasis === netsuiteSalesTax.taxBasis &&
      this.taxRate === netsuiteSalesTax.taxRate &&
      this.taxAmount === netsuiteSalesTax.taxAmount
    ) {
      return true;
    }
    return false;
  }

  /**
   * @private  This method is for internal use only
   */
  dump() {
    console.log('  ***************************');
    console.log('  ** Tax Details...');
    console.log('  ***************************');
    console.log(`    taxDetailsReference: ${this.taxDetailsReference}`);
    console.log(`    lineType: ${this.lineType}`);
    console.log(`    lineName: ${this.lineName}`);
    console.log(`    taxTypeInternalId: ${this.taxTypeInternalId}`);
    console.log(`    taxTypeName: ${this.taxTypeName}`);
    console.log(`    taxCodeInternalId: ${this.taxCodeInternalId}`);
    console.log(`    taxCodeName: ${this.taxCodeName}`);
    console.log(`    calcDetail: ${this.calcDetail}`);
    console.log(`    netAmount: ${this.netAmount}`);
    console.log(`    taxBasis: ${this.taxBasis}`);
    console.log(`    taxRate: ${this.taxRate}`);
    console.log(`    taxAmount: ${this.taxAmount}`);
  }
}

module.exports = NetSuiteSalesTax;
