// @fye/netsuite-client/lib/netsuite-item-fulfillment-shipping.js
/* eslint-disable lines-between-class-members */
/* eslint-disable no-console */
// const debug = require('debug')('netsuite-item-fulfillment-shipping');
const NetSuiteAddress = require('./netsuite-address');

/**
 * NetSuiteItemFulfillmentShipping class
 */
class NetSuiteItemFulfillmentShipping {
  /**
   * Create an instance of the NetSuiteItemFulfillmentShipping class
   * @param  {object} props
   * @param  {object} props.address  Address json object
   * @param  {string} props.method   Method
   * @param  {string} props.status   Status
   */
  constructor({ address, method, status }) {
    this.address = new NetSuiteAddress(address);
    this.method = method;
    this.status = status;
  }

  /**
   * Compare against another NetSuiteItemFulfillmentShipping object to see if they're equal
   * @param {NetSuiteItemFulfillmentShipping} netsuiteItemFulfillmentShipping  The object to compare against
   * @return {boolean} true if they're equal, else returns false
   */
  equals(netsuiteItemFulfillmentShipping) {
    if (
      this.address.equals(netsuiteItemFulfillmentShipping.address) &&
      this.method === netsuiteItemFulfillmentShipping.method &&
      this.status === netsuiteItemFulfillmentShipping.status
    ) {
      return true;
    }
    return false;
  }

  /**
   * Output the properties to the console
   */
  dump() {
    console.log('  ***************************');
    console.log('  ** Shipping Details...');
    console.log('  ***************************');
    this.address.dump();
    console.log(`    method: ${this.method}`);
    console.log(`    status: ${this.status}`);
  }
}

module.exports = NetSuiteItemFulfillmentShipping;
