// @fye/netsuite-client/lib/netsuite-sales-product.js
/* eslint-disable lines-between-class-members */
/* eslint-disable no-console */
// const debug = require('debug')('netsuite-sales-product');
const NetSuiteProduct = require('./netsuite-product');

/**
 * NetSuiteSalesProduct class
 * @extends NetSuiteProduct
 */
class NetSuiteSalesProduct extends NetSuiteProduct {
  /**
   * @private  This constructor is for internal use only
   */
  constructor({
    internalId,
    name,
    description,
    quantity,
    costEstimateType,
    costEstimate,
    rate,
    amount,
    grossAmount,
    taxAmount,
    taxDetailsReference,
  }) {
    super({
      internalId,
      name,
      description,
      quantity,
    });
    this.costEstimateType = costEstimateType;
    this.costEstimate = costEstimate;
    this.rate = rate !== undefined ? rate : 'not available';
    this.amount = amount;
    this.grossAmount = grossAmount;
    this.taxAmount = taxAmount;
    this.taxDetailsReference = taxDetailsReference;
  }

  /**
   * @private  This method is for internal use only
   */
  equals(netsuiteSalesProduct) {
    if (
      super.equals(netsuiteSalesProduct) &&
      this.costEstimateType === netsuiteSalesProduct.costEstimateType &&
      this.costEstimate === netsuiteSalesProduct.costEstimate &&
      this.rate === netsuiteSalesProduct.rate &&
      this.amount === netsuiteSalesProduct.amount &&
      this.grossAmount === netsuiteSalesProduct.grossAmount &&
      this.taxAmount === netsuiteSalesProduct.taxAmount &&
      this.taxDetailsReference === netsuiteSalesProduct.taxDetailsReference
    ) {
      return true;
    }
    return false;
  }

  /**
   * @private  This method is for internal use only
   */
  dump() {
    super.dump();
    console.log(`    costEstimateType: ${this.costEstimateType}`);
    console.log(`    costEstimate: ${this.costEstimate}`);
    console.log(`    rate: ${this.rate}`);
    console.log(`    amount: ${this.amount}`);
    console.log(`    grossAmount: ${this.grossAmount}`);
    console.log(`    taxAmount: ${this.taxAmount}`);
    console.log(`    taxDetailsReference: ${this.taxDetailsReference}`);
  }
}

module.exports = NetSuiteSalesProduct;
