// @fye/netsuite-client/lib/netsuite-item-fulfillment-package.js
/* eslint-disable lines-between-class-members */
/* eslint-disable no-console */
// const debug = require('debug')('netsuite-item-fulfillment-package');

/**
 * NetSuiteItemFulfillmentPackage class
 */
class NetSuiteItemFulfillmentPackage {
  /**
   * Create an instance of the NetSuiteItemFulfillmentPackage class
   * @param  {object} props
   * @param  {number} props.weight          Weight
   * @param  {string} props.description     Description
   * @param  {string} props.trackingNumber  Tracking number
   */
  constructor({ weight, description, trackingNumber }) {
    this.weight = weight;
    this.description = description !== undefined ? description : 'not available';
    this.trackingNumber = trackingNumber;
  }

  /**
   * Compare against another NetSuiteItemFulfillmentPackage object to see if they're equal
   * @param {NetSuiteItemFulfillmentPackage} netsuiteItemFulfillmentPackage  The object to compare against
   * @return {boolean} true if they're equal, else returns false
   */
  // Note: not using trackingNumber value when comparing objects as they may change for each transaction
  equals(netsuiteItemFulfillmentPackage) {
    if (
      this.weight === netsuiteItemFulfillmentPackage.weight &&
      this.description === netsuiteItemFulfillmentPackage.description
    ) {
      return true;
    }
    return false;
  }

  /**
   * Output the properties to the console
   */
  dump() {
    console.log('  ***************************');
    console.log('  ** Package Details...');
    console.log('  ***************************');
    console.log(`    weight: ${this.weight}`);
    console.log(`    description: ${this.description}`);
    console.log(`    trackingNumber: ${this.trackingNumber}`);
  }
}

module.exports = NetSuiteItemFulfillmentPackage;
