// @fye/netsuite-client/lib/netsuite-product.js
/* eslint-disable lines-between-class-members */
/* eslint-disable no-console */
// const debug = require('debug')('netsuite-product');

/**
 * NetSuiteProduct class
 */
class NetSuiteProduct {
  /**
   * @private  This constructor is for internal use only
   */
  constructor({ internalId, name, description, quantity }) {
    this.internalId = internalId;
    this.name = name;
    this.description = description;
    this.quantity = quantity;
  }

  /**
   * @private  This method is for internal use only
   */
  equals(netsuiteProduct) {
    if (
      this.internalId === netsuiteProduct.internalId &&
      this.name === netsuiteProduct.name &&
      this.description === netsuiteProduct.description &&
      this.quantity === netsuiteProduct.quantity
    ) {
      return true;
    }
    return false;
  }

  /**
   * @private  This method is for internal use only
   */
  deepEquals(netsuiteProduct) {
    return this.equals(netsuiteProduct);
  }

  /**
   * @private  This method is for internal use only
   */
  dump() {
    console.log('  ***************************');
    console.log('  ** Product Details...');
    console.log('  ***************************');
    console.log(`    internalId: ${this.internalId}`);
    console.log(`    name: ${this.name}`);
    console.log(`    description: ${this.description}`);
    console.log(`    quantity: ${this.quantity}`);
  }
}

module.exports = NetSuiteProduct;
