// @fye/netsuite-client/lib/netsuite-sales-order-shipping.js
/* eslint-disable lines-between-class-members */
/* eslint-disable no-console */
// const debug = require('debug')('netsuite-sales-order-shipping');
const NetSuiteSalesShipping = require('./netsuite-sales-shipping');

/**
 * NetSuiteSalesOrderShipping class
 * @extends NetSuiteSalesShipping
 */
class NetSuiteSalesOrderShipping extends NetSuiteSalesShipping {
  /**
   * Create an instance of the NetSuiteSalesOrderShipping class
   * @param  {object} props
   * @param  {object} props.address              Address json object
   * @param  {number} props.quantity             Quantity
   * @param  {number} props.amount               Amount
   * @param  {number} props.grossAmount          Gross amount
   * @param  {number} props.taxAmount            Tax amount
   * @param  {string} props.costEstimateType     Cost estimate type
   * @param  {number} props.costEstimate         Cost estimate
   * @param  {string} props.taxDetailsReference  Tax details reference
   * @param  {string} props.date                 Date
   * @param  {string} props.method               Method
   * @param  {string} props.trackingNumber       Tracking number
   * @param  {number} props.quantityBilled       Quantity billed
   */
  constructor({
    address,
    quantity,
    amount,
    grossAmount,
    taxAmount,
    costEstimateType,
    costEstimate,
    taxDetailsReference,
    date,
    method,
    trackingNumber,
    quantityBilled,
  }) {
    super({
      address,
      quantity,
      amount,
      grossAmount,
      taxAmount,
      costEstimateType,
      costEstimate,
      taxDetailsReference,
    });
    this.date = new Date(date);
    this.method = method;
    this.trackingNumber = trackingNumber !== undefined ? trackingNumber : 'not available';
    this.quantityBilled = quantityBilled;
  }

  /**
   * Compare against another NetSuiteSalesOrderShipping object to see if they're equal
   * @param {NetSuiteSalesOrderShipping} netsuiteSalesOrderShipping  The object to compare against
   * @return {boolean} true if they're equal, else returns false
   */
  // Note: not using Date and TrackingNumber values when comparing objects as they may change for each transaction
  equals(netsuiteSalesOrderShipping) {
    if (
      super.equals(netsuiteSalesOrderShipping) &&
      this.method === netsuiteSalesOrderShipping.method &&
      this.quantityBilled === netsuiteSalesOrderShipping.quantityBilled
    ) {
      return true;
    }
    return false;
  }

  /**
   * Output the properties to the console
   */
  dump() {
    super.dump();
    console.log(`    date: ${this.date}`);
    console.log(`    method: ${this.method}`);
    console.log(`    trackingNumber: ${this.trackingNumber}`);
    console.log(`    quantityBilled: ${this.quantityBilled}`);
  }
}

module.exports = NetSuiteSalesOrderShipping;
