// @fye/netsuite-client/lib/netsuite-sales-shipping.js
/* eslint-disable lines-between-class-members */
/* eslint-disable no-console */
// const debug = require('debug')('netsuite-sales-shipping');
const NetSuiteAddress = require('./netsuite-address');

/**
 * NetSuiteSalesShipping class
 */
class NetSuiteSalesShipping {
  /**
   * @private  This constructor is for internal use only
   */
  constructor({
    address,
    quantity,
    amount,
    grossAmount,
    taxAmount,
    costEstimateType,
    costEstimate,
    taxDetailsReference,
  }) {
    this.address = new NetSuiteAddress(address);
    this.quantity = quantity;
    this.amount = amount;
    this.grossAmount = grossAmount;
    this.taxAmount = taxAmount;
    this.costEstimateType = costEstimateType;
    this.costEstimate = costEstimate;
    this.taxDetailsReference = taxDetailsReference;
  }

  /**
   * @private  This method is for internal use only
   */
  equals(netsuiteSalesShipping) {
    if (
      this.address.equals(netsuiteSalesShipping.address) &&
      this.quantity === netsuiteSalesShipping.quantity &&
      this.amount === netsuiteSalesShipping.amount &&
      this.grossAmount === netsuiteSalesShipping.grossAmount &&
      this.taxAmount === netsuiteSalesShipping.taxAmount &&
      this.costEstimateType === netsuiteSalesShipping.costEstimateType &&
      this.costEstimate === netsuiteSalesShipping.costEstimate &&
      this.taxDetailsReference === netsuiteSalesShipping.taxDetailsReference
    ) {
      return true;
    }
    return false;
  }

  /**
   * @private  This method is for internal use only
   */
  dump() {
    console.log('  ***************************');
    console.log('  ** Shipping Details...');
    console.log('  ***************************');
    this.address.dump();
    console.log(`    quantity: ${this.quantity}`);
    console.log(`    amount: ${this.amount}`);
    console.log(`    grossAmount: ${this.grossAmount}`);
    console.log(`    taxAmount: ${this.taxAmount}`);
    console.log(`    costEstimateType: ${this.costEstimateType}`);
    console.log(`    costEstimate: ${this.costEstimate}`);
    console.log(`    taxDetailsReference: ${this.taxDetailsReference}`);
  }
}

module.exports = NetSuiteSalesShipping;
